Mathlib/AlgebraicGeometry/EllipticCurve/NormalForms.lean

```diff
@@ -0,0 +1,661 @@
+  linear_combination (-W.a₂ - W.a₁ ^ 2) * CharP.cast_eq_zero R 3
+
+theorem toShortNFOfCharThree_spec (hb₂ : W.b₂ = 0) :
+    (W.variableChange W.toShortNFOfCharThree).IsShortNF := by
+  have h : (2 : R) * 2 = 1 := by linear_combination CharP.cast_eq_zero R 3
+  letI : Invertible (2 : R) := ⟨2, h, h⟩
+  have H := W.toCharNeTwoNF_spec
+  exact ⟨H.a₁, hb₂ ▸ W.toShortNFOfCharThree_a₂, H.a₃⟩
+
+section
+
+variable (W : WeierstrassCurve F)
+
+theorem toCharThreeNF_spec_of_b₂_ne_zero (hb₂ : W.b₂ ≠ 0) :
+    (W.variableChange W.toCharThreeNF).IsCharThreeJNeZeroNF := by
+  have h : (2 : F) * 2 = 1 := by linear_combination CharP.cast_eq_zero F 3
+  letI : Invertible (2 : F) := ⟨2, h, h⟩
+  rw [toCharThreeNF]; erw [variableChange_comp]
+  rw [← toShortNFOfCharThree_a₂] at hb₂
+  set W' := W.variableChange W.toShortNFOfCharThree
+  have H : W'.IsCharNeTwoNF := W.toCharNeTwoNF_spec
+  constructor
+  · simp [H.a₁]
+  · simp [H.a₁, H.a₃]
+  · field_simp
+    linear_combination (W'.a₄ * W'.a₂ ^ 2 + W'.a₄ ^ 2) * CharP.cast_eq_zero F 3
+
+theorem toCharThreeNF_spec_of_b₂_zero (hb₂ : W.b₂ = 0) :
+    (W.variableChange W.toCharThreeNF).IsShortNF := by
+  rw [toCharThreeNF, toShortNFOfCharThree_a₂, hb₂, div_zero]
+  erw [one_mul]
+  exact W.toShortNFOfCharThree_spec hb₂
+
+theorem toCharThreeNF_spec : (W.variableChange W.toCharThreeNF).IsCharThreeNF := by
+  by_cases hb₂ : W.b₂ = 0
+  · exact Or.inr (W.toCharThreeNF_spec_of_b₂_zero hb₂)
+  · exact Or.inl (W.toCharThreeNF_spec_of_b₂_ne_zero hb₂)
+
+theorem exists_variableChange_isCharThreeNF :
+    ∃ C : VariableChange F, (W.variableChange C).IsCharThreeNF :=
+  ⟨_, W.toCharThreeNF_spec⟩
+
+end
+
+end VariableChange
+
+/-! ### Normal forms of characteristic = 2 and j ≠ 0 -/
+
+/-- A `WeierstrassCurve` is in normal form of characteristic = 2 and j ≠ 0, if its $a_1 = 1$,
+$a_3, a_4 = 0$. In other words it is $Y^2 + XY = X^3 + a_2X^2 + a_6$. -/
+@[mk_iff]
+structure IsCharTwoJNeZeroNF : Prop where
+  a₁ : W.a₁ = 1
+  a₃ : W.a₃ = 0
+  a₄ : W.a₄ = 0
+
+-- TODO: move to suitable location ???
+theorem b₂_of_char_two [CharP R 2] : W.b₂ = W.a₁ ^ 2 := by
+  rw [b₂]
+  linear_combination 2 * W.a₂ * CharP.cast_eq_zero R 2
+
+-- TODO: move to suitable location ???
+theorem b₄_of_char_two [CharP R 2] : W.b₄ = W.a₁ * W.a₃ := by
+  rw [b₄]
+  linear_combination W.a₄ * CharP.cast_eq_zero R 2
+
+-- TODO: move to suitable location ???
+theorem b₆_of_char_two [CharP R 2] : W.b₆ = W.a₃ ^ 2 := by
+  rw [b₆]
+  linear_combination 2 * W.a₆ * CharP.cast_eq_zero R 2
+
+-- TODO: move to suitable location ???
+theorem c₄_of_char_two [CharP R 2] : W.c₄ = W.a₁ ^ 4 := by
+  rw [c₄, b₂_of_char_two]
+  linear_combination -12 * W.b₄ * CharP.cast_eq_zero R 2
+
+-- TODO: move to suitable location ???
+theorem c₆_of_char_two [CharP R 2] : W.c₆ = -W.a₁ ^ 6 := by
+  rw [c₆, b₂_of_char_two]
+  linear_combination (18 * W.a₁ ^ 2 * W.b₄ - 108 * W.b₆) * CharP.cast_eq_zero R 2
```

Comment:
I guess you want these in `Weierstrass.lean`? Maybe add it for Delta/j and char 3 too if they're nice enough.